nixos/modules/installer/tools/nixos-rebuild.sh

```diff
@@ -80,6 +90,87 @@ while [ "$#" -gt 0 ]; do
     esac
 done
 
+
+if [ -z "$buildHost" -a -n "$targetHost" ]; then
+  buildHost="$targetHost"
+fi
+if [ "$targetHost" = localhost ]; then
+  targetHost=
+fi
+if [ "$buildHost" = localhost ]; then
+  buildHost=
+fi
+
+
+buildHostCmd() {
+  if [ -z "$buildHost" ]; then
+    "$@"
+  else
+    ssh "$buildHost" PATH="$tmpDir/nix/bin:$PATH" "$@"
+  fi
+}
+
+targetHostCmd() {
+  if [ -z "$targetHost" ]; then
+    "$@"
+  else
+    ssh "$targetHost" "$@"
+  fi
+}
+
+copyToTarget() {
+  if ! [ "$targetHost" = "$buildHost" ]; then
+    if [ -z "$targetHost" ]; then
+      nix-copy-closure --from "$buildHost" "$1"
+    else
+      buildHostCmd nix-copy-closure --to "$targetHost" "$1"
+    fi
+  fi
+}
+
+nixBuild() {
+  if [ -z "$buildHost" ]; then
+    nix-build "$@"
+  else
+    local instArgs=()
+    local buildArgs=()
+
+    while [ "$#" -gt 0 ]; do
+      local i="$1"; shift 1
+      case "$i" in
+        -o)
+          local out="$1"; shift 1
+          buildArgs+=("--add-root" "$out" "--indirect")
+          ;;
+        -A)
+          local j="$1"; shift 1
+          instArgs+=("$i" "$j")
+          ;;
+        -I)
+          # We don't want this in buildArgs
+          shift 1
+          ;;
+        "<"*) # nix paths
+          instArgs+=("$i")
+          ;;
+        *)
+          buildArgs+=("$i")
+          ;;
+      esac
+    done
+
+    local drv="$(nix-instantiate "${instArgs[@]}" "${extraBuildFlags[@]}")"
+    if [ -a "$drv" ]; then
+      nix-copy-closure --to "$buildHost" "$drv"
+      ssh "$buildHost" nix-store -r "$drv" "${buildArgs[@]}"
```

Comment:
I will recommend to use a ssh master connection, such that the user does not have to wonder why he has to login twice.